ledger/src/main/scala/co/topl/ledger/interpreters/Mempool.scala

```
@@ -91,24 +95,46 @@ object Mempool {
           .map(_.transactionIds.map(t => t: TypedIdentifier).toList)
           .flatMap(_.traverse(fetchTransaction(_).flatMap(addTransactionWithDefaultExpiration)))
           .as(state)
-      eventSourcedState <- EventSourcedState.OfTree.make[F, State[F], TypedIdentifier](
-        state.pure[F],
-        currentBlockId,
-        applyEvent = applyBlock,
-        unapplyEvent = unapplyBlock,
-        parentChildTree,
-        currentEventChanged
+      eventSourcedState <- Resource.eval(
+        EventSourcedState.OfTree.make[F, State[F], TypedIdentifier](
+          state.pure[F],
+          currentBlockId,
+          applyEvent = applyBlock,
+          unapplyEvent = unapplyBlock,
+          parentChildTree,
+          currentEventChanged
+        )
       )
+      finalizing <- Resource.make(Deferred[F, Unit])(_.complete(()).void)
     } yield new MempoolAlgebra[F] {
 
       def read(blockId: TypedIdentifier): F[Set[TypedIdentifier]] =
-        eventSourcedState.stateAt(blockId).flatMap(_.get).map(_.keySet)
+        whenNotTerminated(
+          eventSourcedState.stateAt(blockId).flatMap(_.get).map(_.keySet)
+        )
 
       // TODO: Check for double-spends along current canonical chain?
       def add(transactionId: TypedIdentifier): F[Unit] =
-        fetchTransaction(transactionId).flatMap(addTransactionWithDefaultExpiration)
+        whenNotTerminated(
+          fetchTransaction(transactionId).flatMap(addTransactionWithDefaultExpiration)
+        )
 
       def remove(transactionId: TypedIdentifier): F[Unit] =
-        state.update(_ - transactionId)
+        whenNotTerminated(
+          state
+            .getAndUpdate(_ - transactionId)
+            .flatTap(entries =>
+              OptionT.fromOption[F](entries.get(transactionId)).foldF(Applicative[F].unit)(_.expirationFiber.cancel)
```

LGTM, 1 question,
- fiber cancelation here, is because the resource release cancel, will be invoked at the end?

---

When removing an entry from the mempool, it's not necessary to keep around the background fiber associated with the entry.  We can eagerly cancel it instead.  The previous behavior was to allow the callback to still take place; it would just be silently ignored.  This change is just a slight optimization to reduce the number of unnecessary background fibers.